# 8D Structure
<!-- 2026.03.21: DAJones file created
-->
Referenced in [8D Investigations](8D%20Investigations.ipynb)




In [1]:
import CliffordTools as CT
import util

dimensions = 5
elements = 1 << dimensions
pseudo_scalar = elements - 1
Grade = [0]
while len(Grade) < elements:
    Grade += [ g+1 for g in Grade ]

#ST = CT.SignTable( list(range(elements)) )
#Ref = [ CT.SignRowToIndex( r ) for r in ST ]

R = list(range(elements))
print( "Xor Test R ", util.XorTest( R ) )
Ref = [ CT.SignRowToIndex( CT.SignRow(r,R) ) for r in R  ]

N = list(range(elements >> 1))
msb = elements >> 1
Parity = [ (1&Grade[n])<<(dimensions-1) for n in N ]
Even = [ n + Parity[n] for n in N ]
Odd = [ Even[n] ^ msb for n in N ]
for n in N:
    even = Even[n]
    odd = msb ^ even
    print( f"{n:>4d} {n:08b}   || {even:>4d} {even:08b}   | {Ref[even]:>4d} {Ref[even]:08b}   || {odd:>4d} {odd:08b}   | {Ref[odd]:>4d} {Ref[odd]:08b}" )

RefEven = [ Ref[even] for even in Even ] 
print( "Xor Test Even ", util.XorTest( Even ) )
print( "Xor Test RefEven ", util.XorTest( RefEven ) )

n = 1
while n < msb:
    even = Even[n]
    odd = msb ^ even
    print( f"{n:>4d} {n:08b}   || {even:>4d} {even:08b}   | {Ref[even]:>4d} {Ref[even]:08b}   || {odd:>4d} {odd:08b}   | {Ref[odd]:>4d} {Ref[odd]:08b}" )
    n <<= 1


the correct version is imported
Xor Test R  True
   0 00000000   ||    0 00000000   |    0 00000000   ||   16 00010000   |    0 00000000
   1 00000001   ||   17 00010001   |   30 00011110   ||    1 00000001   |   30 00011110
   2 00000010   ||   18 00010010   |   28 00011100   ||    2 00000010   |   28 00011100
   3 00000011   ||    3 00000011   |    2 00000010   ||   19 00010011   |    2 00000010
   4 00000100   ||   20 00010100   |   24 00011000   ||    4 00000100   |   24 00011000
   5 00000101   ||    5 00000101   |    6 00000110   ||   21 00010101   |    6 00000110
   6 00000110   ||    6 00000110   |    4 00000100   ||   22 00010110   |    4 00000100
   7 00000111   ||   23 00010111   |   26 00011010   ||    7 00000111   |   26 00011010
   8 00001000   ||   24 00011000   |   16 00010000   ||    8 00001000   |   16 00010000
   9 00001001   ||    9 00001001   |   14 00001110   ||   25 00011001   |   14 00001110
  10 00001010   ||   10 00001010   |   12 00001100   ||   26 00011010  

In [6]:

# the same code as in the previous cell except `SignCol` replaces `SignRow`
# these tables relate to a commutated multiplication rule.

R = list(range(elements))
print( "Xor Test R ", util.XorTest( R ) )
Ref = [ CT.SignRowToIndex( CT.SignCol(r,R) ) for r in R  ]

N = list(range(elements >> 1))
msb = elements >> 1
Parity = [ (1&Grade[n])<<(dimensions-1) for n in N ]
Even = [ n + Parity[n] for n in N ]
Odd = [ Even[n] ^ 1 for n in N ]
for n in N:
    even = Even[n]
    odd = Odd[n]
    print( f"{n:>4d} {n:08b}   || {even:>4d} {even:08b}   | {Ref[even]:>4d} {Ref[even]:08b}   || {odd:>4d} {odd:08b}   | {Ref[odd]:>4d} {Ref[odd]:08b}" )

RefEven = [ Ref[even] for even in Even ] 
print( "Xor Test Even ", util.XorTest( Even ) )
print( "Xor Test RefEven ", util.XorTest( RefEven ) )

n = 1
while n < msb:
    even = Even[n]
    odd = Odd[n]
    print( f"{n:>4d} {n:08b}   || {even:>4d} {even:08b}   | {Ref[even]:>4d} {Ref[even]:08b}   || {odd:>4d} {odd:08b}   | {Ref[odd]:>4d} {Ref[odd]:08b}" )
    n <<= 1

Xor Test R  True
   0 00000000   ||    0 00000000   |    0 00000000   ||    1 00000001   |    0 00000000
   1 00000001   ||   17 00010001   |   15 00001111   ||   16 00010000   |   15 00001111
   2 00000010   ||   18 00010010   |   14 00001110   ||   19 00010011   |   14 00001110
   3 00000011   ||    3 00000011   |    1 00000001   ||    2 00000010   |    1 00000001
   4 00000100   ||   20 00010100   |   12 00001100   ||   21 00010101   |   12 00001100
   5 00000101   ||    5 00000101   |    3 00000011   ||    4 00000100   |    3 00000011
   6 00000110   ||    6 00000110   |    2 00000010   ||    7 00000111   |    2 00000010
   7 00000111   ||   23 00010111   |   13 00001101   ||   22 00010110   |   13 00001101
   8 00001000   ||   24 00011000   |    8 00001000   ||   25 00011001   |    8 00001000
   9 00001001   ||    9 00001001   |    7 00000111   ||    8 00001000   |    7 00000111
  10 00001010   ||   10 00001010   |    6 00000110   ||   11 00001011   |    6 00000110
  11 00001011  




NORMAL
|    element     |    even basis   |    ref basis    |   odd "basis"   |    odd "ref"    |
|   :--------:   |   :---------:   |   :---------:   |   :---------:   |   :---------:   |
|   1 00000001   |   17 00010001   |   30 00011110   |    1 00000001   |   30 00011110   |
|   2 00000010   |   18 00010010   |   28 00011100   |    2 00000010   |   28 00011100   |
|   4 00000100   |   20 00010100   |   24 00011000   |    4 00000100   |   24 00011000   |
|   8 00001000   |   24 00011000   |   16 00010000   |    8 00001000   |   16 00010000   |
   
COMMUTATED   
|    element     |    even basis   |    ref basis    |   odd "basis"   |    odd "ref"    |
|   ----------   |   -----------   |   -----------   |   -----------   |   -----------   |
|   1 00000001   |   17 00010001   |   15 00001111   |    1 00000001   |    0 00000000   |
|   2 00000010   |   18 00010010   |   14 00001110   |    2 00000010   |    1 00000001   |
|   4 00000100   |   20 00010100   |   12 00001100   |    4 00000100   |    3 00000011   |
|   8 00001000   |   24 00011000   |    8 00001000   |    8 00001000   |    7 00000111   |